Crash when adding a self hosted site

Originally from [here](https://github.com/wordpress-mobile/WordPress-Android/pull/4455#issuecomment-242238265). In the Example App:

* Log into .com account with a Jetpack site
* Log into that Jetpack site as self-hosted
* Site -> Fetch All sites
* 💥 

```
Caused by: android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: SiteModel.SITE_ID, SiteModel.URL (code 2067)
    at android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
    at android.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:732)
    at android.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:754)
    at android.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:64)
    at android.database.sqlite.SQLiteDatabase.updateWithOnConflict(SQLiteDatabase.java:1576)
    at android.database.sqlite.SQLiteDatabase.update(SQLiteDatabase.java:1522)
    at com.yarolegovich.wellsql.UpdateQuery.execute(UpdateQuery.java:79)
    at org.wordpress.android.fluxc.persistence.SiteSqlUtils.insertOrUpdateSite(SiteSqlUtils.java:87)
    at org.wordpress.android.fluxc.store.SiteStore.updateSite(SiteStore.java:566)
    at org.wordpress.android.fluxc.store.SiteStore.onAction(SiteStore.java:502)
    at java.lang.reflect.Method.invoke(Native Method)
    at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:485)
    at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:479) 
    at org.greenrobot.eventbus.AsyncPoster.run(AsyncPoster.java:46) 
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
    at java.lang.Thread.run(Thread.java:818) 
```

--- a/example/src/main/java/org/wordpress/android/fluxc/example/MainFragment.java
+++ b/example/src/main/java/org/wordpress/android/fluxc/example/MainFragment.java
@@ -327,7 +327,8 @@ public class MainFragment extends Fragment {
         }
         if (mSiteStore.hasSite()) {
             SiteModel firstSite = mSiteStore.getSites().get(0);
-            prependToLog("First site name: " + firstSite.getName() + " - Total sites: " + mSiteStore.getSitesCount());
+            prependToLog("First site name: " + firstSite.getName() + " - Total sites: " + mSiteStore.getSitesCount()
+                         + " - rowsAffected: " + event.rowsAffected);
         }
     }
 

--- a/example/src/main/java/org/wordpress/android/fluxc/example/SitesFragment.java
+++ b/example/src/main/java/org/wordpress/android/fluxc/example/SitesFragment.java
@@ -60,6 +60,15 @@ public class SitesFragment extends Fragment {
             }
         });
 
+        view.findViewById(R.id.update_all_sites).setOnClickListener(new OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                for (SiteModel site: mSiteStore.getSites()) {
+                    mDispatcher.dispatch(SiteActionBuilder.newFetchSiteAction(site));
+                }
+            }
+        });
+
         view.findViewById(R.id.log_sites).setOnClickListener(new OnClickListener() {
             @Override
             public void onClick(View v) {

--- a/example/src/main/java/org/wordpress/android/fluxc/example/SitesFragment.java
+++ b/example/src/main/java/org/wordpress/android/fluxc/example/SitesFragment.java
@@ -122,16 +122,16 @@ public class SitesFragment extends Fragment {
 
     @Subscribe(threadMode = ThreadMode.MAIN)
     public void onSiteChanged(OnSiteChanged event) {
-        prependToLog("OnSiteChanged");
         if (event.isError()) {
+            prependToLog("SiteChanged error: " + event.error.type);
             AppLog.e(T.TESTS, "SiteChanged error: " + event.error.type);
-            return;
+        } else {
+            prependToLog("SiteChanged: rowsAffected = " + event.rowsAffected);
         }
     }
 
     @Subscribe(threadMode = ThreadMode.MAIN)
     public void onNewSiteCreated(OnNewSiteCreated event) {
-        prependToLog("OnNewSiteCreated");
         String message = event.dryRun ? "validated" : "created";
         if (event.isError()) {
             prependToLog("New site " + message + ": error: " + event.error.type + " - " + event.error.message);

--- a/example/src/main/res/layout/fragment_sites.xml
+++ b/example/src/main/res/layout/fragment_sites.xml
@@ -18,6 +18,12 @@
         android:text="Fetch first site" />
 
     <Button
+        android:id="@+id/update_all_sites"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="Fetch all sites" />
+
+    <Button
         android:id="@+id/log_sites"
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"

--- a/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
+++ b/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
@@ -25,7 +25,6 @@ import java.util.List;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
@@ -33,6 +32,7 @@ import static org.wordpress.android.fluxc.utils.SiteUtils.generateJetpackSite;
 import static org.wordpress.android.fluxc.utils.SiteUtils.generateJetpackSiteOverRestOnly;
 import static org.wordpress.android.fluxc.utils.SiteUtils.generatePostFormats;
 import static org.wordpress.android.fluxc.utils.SiteUtils.generateSelfHostedNonJPSite;
+import static org.wordpress.android.fluxc.utils.SiteUtils.generateSelfHostedSiteFutureJetpack;
 import static org.wordpress.android.fluxc.utils.SiteUtils.generateWPComSite;
 
 @RunWith(RobolectricTestRunner.class)
@@ -272,13 +272,52 @@ public class SiteStoreUnitTest {
         SiteSqlUtils.insertOrUpdateSite(wpComSite);
         SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverSelfHosted);
         SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverRestOnly);
+        SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverRestOnly);
 
         List<SiteModel> wpComSites = SiteSqlUtils.getAllWPComSites();
 
-        assertEquals(2, wpComSites.size());
-        for (SiteModel site : wpComSites) {
-            assertNotEquals(jetpackSiteOverSelfHosted.getId(), site.getId());
+        assertEquals(3, wpComSites.size());
+    }
+
+    @Test
+    public void testInsertDuplicateSites() throws DuplicateSiteException {
+        SiteModel futureJetpack = generateSelfHostedSiteFutureJetpack();
+        SiteModel jetpack = generateJetpackSiteOverRestOnly();
+
+        // Insert a self hosted site that will later be converted to Jetpack
+        SiteSqlUtils.insertOrUpdateSite(futureJetpack);
+        // Insert the same site but Jetpack powered this time
+        SiteSqlUtils.insertOrUpdateSite(jetpack);
+
+        // Previous site should be converted to a Jetpack site and we should see only one site
+        int sitesCount = WellSql.select(SiteModel.class).getAsCursor().getCount();
+        assertEquals(1, sitesCount);
+
+        List<SiteModel> wpComSites  = SiteSqlUtils.getAllWPComSites();
+        assertEquals(1, wpComSites.size());
+        assertEquals(jetpack.getSiteId(), wpComSites.get(0).getSiteId());
+        assertTrue(wpComSites.get(0).isJetpack());
+        assertTrue(wpComSites.get(0).isWPCom());
+    }
+
+    @Test
+    public void testInsertDuplicateSitesError() throws DuplicateSiteException {
+        SiteModel futureJetpack = generateSelfHostedSiteFutureJetpack();
+        SiteModel jetpack = generateJetpackSiteOverRestOnly();
+
+        // Insert a Jetpack powered site
+        SiteSqlUtils.insertOrUpdateSite(jetpack);
+        boolean duplicate = false;
+        try {
+            // Insert the same site but via self hosted this time (this should fail)
+            SiteSqlUtils.insertOrUpdateSite(futureJetpack);
+        } catch (DuplicateSiteException e) {
+            // Caught !
+            duplicate = true;
         }
+        assertTrue(duplicate);
+        int sitesCount = WellSql.select(SiteModel.class).getAsCursor().getCount();
+        assertEquals(1, sitesCount);
     }
 
     @Test

--- a/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
+++ b/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
@@ -25,10 +25,11 @@ import java.util.List;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
-import static org.wordpress.android.fluxc.utils.SiteUtils.generateJetpackSite;
+import static org.wordpress.android.fluxc.utils.SiteUtils.generateJetpackSiteOverXMLRPC;
 import static org.wordpress.android.fluxc.utils.SiteUtils.generateJetpackSiteOverRestOnly;
 import static org.wordpress.android.fluxc.utils.SiteUtils.generatePostFormats;
 import static org.wordpress.android.fluxc.utils.SiteUtils.generateSelfHostedNonJPSite;
@@ -100,8 +101,8 @@ public class SiteStoreUnitTest {
         assertEquals(1, mSiteStore.getSelfHostedSitesCount());
 
         // Test counts with one .COM, one self-hosted and one Jetpack site
-        SiteModel jetpackSite = generateJetpackSite();
-        SiteSqlUtils.insertOrUpdateSite(jetpackSite);
+        SiteModel jetpackSiteOverXMLRPC = generateJetpackSiteOverXMLRPC();
+        SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverXMLRPC);
 
         assertTrue(mSiteStore.hasSite());
         assertTrue(mSiteStore.hasWPComSite());
@@ -109,8 +110,8 @@ public class SiteStoreUnitTest {
         assertTrue(mSiteStore.hasJetpackSite());
 
         assertEquals(3, mSiteStore.getSitesCount());
-        assertEquals(2, mSiteStore.getWPComSitesCount());
-        assertEquals(1, mSiteStore.getSelfHostedSitesCount());
+        assertEquals(1, mSiteStore.getWPComSitesCount());
+        assertEquals(2, mSiteStore.getSelfHostedSitesCount());
         assertEquals(1, mSiteStore.getJetpackSitesCount());
     }
 
@@ -124,7 +125,7 @@ public class SiteStoreUnitTest {
         assertTrue(mSiteStore.hasSelfHostedSiteWithSiteIdAndXmlRpcUrl(selfHostedSite.getSelfHostedSiteId(),
                 selfHostedSite.getXmlRpcUrl()));
 
-        SiteModel jetpackSite = generateJetpackSite();
+        SiteModel jetpackSite = generateJetpackSiteOverXMLRPC();
         SiteSqlUtils.insertOrUpdateSite(jetpackSite);
 
         assertTrue(mSiteStore.hasSelfHostedSiteWithSiteIdAndXmlRpcUrl(jetpackSite.getSelfHostedSiteId(),
@@ -140,7 +141,7 @@ public class SiteStoreUnitTest {
 
         assertTrue(mSiteStore.hasWPComOrJetpackSiteWithSiteId(wpComSite.getSiteId()));
 
-        SiteModel jetpackSite = generateJetpackSite();
+        SiteModel jetpackSite = generateJetpackSiteOverXMLRPC();
         SiteSqlUtils.insertOrUpdateSite(jetpackSite);
 
         // hasWPComOrJetpackSiteWithSiteId() should be able to locate a Jetpack site with either the site id or the
@@ -210,7 +211,7 @@ public class SiteStoreUnitTest {
 
         SiteModel selfHostedSite = generateSelfHostedNonJPSite();
         SiteModel wpComSite = generateWPComSite();
-        SiteModel jetpackSite = generateJetpackSite();
+        SiteModel jetpackSite = generateJetpackSiteOverXMLRPC();
         SiteSqlUtils.insertOrUpdateSite(selfHostedSite);
         SiteSqlUtils.insertOrUpdateSite(wpComSite);
         SiteSqlUtils.insertOrUpdateSite(jetpackSite);
@@ -239,7 +240,7 @@ public class SiteStoreUnitTest {
 
         SiteModel selfHostedSite = generateSelfHostedNonJPSite();
         SiteModel wpComSite = generateWPComSite();
-        SiteModel jetpackSite = generateJetpackSite();
+        SiteModel jetpackSite = generateJetpackSiteOverXMLRPC();
         SiteSqlUtils.insertOrUpdateSite(selfHostedSite);
         SiteSqlUtils.insertOrUpdateSite(wpComSite);
         SiteSqlUtils.insertOrUpdateSite(jetpackSite);
@@ -266,17 +267,19 @@ public class SiteStoreUnitTest {
     @Test
     public void testGetWPComSites() throws DuplicateSiteException {
         SiteModel wpComSite = generateWPComSite();
-        SiteModel jetpackSiteOverSelfHosted = generateJetpackSite();
+        SiteModel jetpackSiteOverXMLRPC = generateJetpackSiteOverXMLRPC();
         SiteModel jetpackSiteOverRestOnly = generateJetpackSiteOverRestOnly();
 
         SiteSqlUtils.insertOrUpdateSite(wpComSite);
-        SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverSelfHosted);
-        SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverRestOnly);
+        SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverXMLRPC);
         SiteSqlUtils.insertOrUpdateSite(jetpackSiteOverRestOnly);
 
         List<SiteModel> wpComSites = SiteSqlUtils.getAllWPComSites();
 
-        assertEquals(3, wpComSites.size());
+        assertEquals(2, wpComSites.size());
+        for (SiteModel site : wpComSites) {
+            assertNotEquals(jetpackSiteOverXMLRPC.getId(), site.getId());
+        }
     }
 
     @Test

--- a/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
+++ b/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
@@ -108,8 +108,8 @@ public class SiteStoreUnitTest {
         assertTrue(mSiteStore.hasJetpackSite());
 
         assertEquals(3, mSiteStore.getSitesCount());
-        assertEquals(1, mSiteStore.getWPComSitesCount());
-        assertEquals(2, mSiteStore.getSelfHostedSitesCount());
+        assertEquals(2, mSiteStore.getWPComSitesCount());
+        assertEquals(1, mSiteStore.getSelfHostedSitesCount());
         assertEquals(1, mSiteStore.getJetpackSitesCount());
     }
 

--- a/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
+++ b/example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java
@@ -17,6 +17,7 @@ import org.wordpress.android.fluxc.model.SiteModel;
 import org.wordpress.android.fluxc.network.rest.wpcom.site.SiteRestClient;
 import org.wordpress.android.fluxc.network.xmlrpc.site.SiteXMLRPCClient;
 import org.wordpress.android.fluxc.persistence.SiteSqlUtils;
+import org.wordpress.android.fluxc.persistence.SiteSqlUtils.DuplicateSiteException;
 import org.wordpress.android.fluxc.persistence.WellSqlConfig;
 import org.wordpress.android.fluxc.store.SiteStore;
 
@@ -60,7 +61,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testInsertOrUpdateSite() {
+    public void testInsertOrUpdateSite() throws DuplicateSiteException {
         SiteModel site = generateWPComSite();
         SiteSqlUtils.insertOrUpdateSite(site);
 
@@ -69,7 +70,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testHasSiteAndgetCountMethods() {
+    public void testHasSiteAndgetCountMethods() throws DuplicateSiteException {
         assertFalse(mSiteStore.hasSite());
         assertTrue(mSiteStore.getSites().isEmpty());
 
@@ -114,7 +115,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testHasSiteWithSiteIdAndXmlRpcUrl() {
+    public void testHasSiteWithSiteIdAndXmlRpcUrl() throws DuplicateSiteException {
         assertFalse(mSiteStore.hasSelfHostedSiteWithSiteIdAndXmlRpcUrl(124, ""));
 
         SiteModel selfHostedSite = generateSelfHostedNonJPSite();
@@ -131,7 +132,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testHasWPComOrJetpackSiteWithSiteId() {
+    public void testHasWPComOrJetpackSiteWithSiteId() throws DuplicateSiteException {
         assertFalse(mSiteStore.hasWPComOrJetpackSiteWithSiteId(673));
 
         SiteModel wpComSite = generateWPComSite();
@@ -149,7 +150,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testWPComSiteVisibility() {
+    public void testWPComSiteVisibility() throws DuplicateSiteException {
         // Should not cause any errors
         mSiteStore.isWPComSiteVisibleByLocalId(45);
         SiteSqlUtils.setSiteVisibility(null, true);
@@ -173,7 +174,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testSetAllWPComSitesVisibility() {
+    public void testSetAllWPComSitesVisibility() throws DuplicateSiteException {
         SiteModel selfHostedNonJPSite = generateSelfHostedNonJPSite();
         SiteSqlUtils.insertOrUpdateSite(selfHostedNonJPSite);
 
@@ -202,7 +203,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testGetIdForIdMethods() {
+    public void testGetIdForIdMethods() throws DuplicateSiteException {
         assertEquals(0, mSiteStore.getLocalIdForRemoteSiteId(555));
         assertEquals(0, mSiteStore.getLocalIdForSelfHostedSiteIdAndXmlRpcUrl(2626, ""));
         assertEquals(0, mSiteStore.getSiteIdForLocalId(5577));
@@ -233,7 +234,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testGetSiteBySiteId() {
+    public void testGetSiteBySiteId() throws DuplicateSiteException {
         assertNull(mSiteStore.getSiteBySiteId(555));
 
         SiteModel selfHostedSite = generateSelfHostedNonJPSite();
@@ -249,7 +250,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testDeleteSite() {
+    public void testDeleteSite() throws DuplicateSiteException {
         SiteModel wpComSite = generateWPComSite();
 
         // Should not cause any errors
@@ -263,7 +264,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testGetWPComSites() {
+    public void testGetWPComSites() throws DuplicateSiteException {
         SiteModel wpComSite = generateWPComSite();
         SiteModel jetpackSiteOverSelfHosted = generateJetpackSite();
         SiteModel jetpackSiteOverRestOnly = generateJetpackSiteOverRestOnly();
@@ -281,7 +282,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testGetPostFormats() {
+    public void testGetPostFormats() throws DuplicateSiteException {
         SiteModel site = generateWPComSite();
         SiteSqlUtils.insertOrUpdateSite(site);
 
@@ -297,7 +298,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testSearchSitesByNameMatching() {
+    public void testSearchSitesByNameMatching() throws DuplicateSiteException {
         SiteModel wpComSite1 = generateWPComSite();
         wpComSite1.setName("Doctor Emmet Brown Homepage");
         SiteModel wpComSite2 = generateWPComSite();
@@ -317,7 +318,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testSearchSitesByNameOrUrlMatching() {
+    public void testSearchSitesByNameOrUrlMatching() throws DuplicateSiteException {
         SiteModel wpComSite1 = generateWPComSite();
         wpComSite1.setName("Doctor Emmet Brown Homepage");
         SiteModel wpComSite2 = generateWPComSite();
@@ -337,7 +338,7 @@ public class SiteStoreUnitTest {
     }
 
     @Test
-    public void testSearchWPComSitesByNameOrUrlMatching() {
+    public void testSearchWPComSitesByNameOrUrlMatching() throws DuplicateSiteException {
         SiteModel wpComSite1 = generateWPComSite();
         wpComSite1.setName("Doctor Emmet Brown Homepage");
         SiteModel wpComSite2 = generateWPComSite();

--- a/example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java
+++ b/example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java
@@ -29,7 +29,7 @@ public class SiteUtils {
         SiteModel example = new SiteModel();
         example.setSiteId(982);
         example.setSelfHostedSiteId(8);
-        example.setIsWPCom(false);
+        example.setIsWPCom(true);
         example.setIsJetpack(true);
         example.setIsVisible(true);
         example.setXmlRpcUrl("http://jetpack.url/xmlrpc.php");
@@ -39,7 +39,7 @@ public class SiteUtils {
     public static SiteModel generateJetpackSiteOverRestOnly() {
         SiteModel example = new SiteModel();
         example.setSiteId(5623);
-        example.setIsWPCom(false);
+        example.setIsWPCom(true);
         example.setIsJetpack(true);
         example.setIsVisible(true);
         example.setXmlRpcUrl("http://jetpack.url/xmlrpc.php");

--- a/example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java
+++ b/example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java
@@ -42,7 +42,17 @@ public class SiteUtils {
         example.setIsWPCom(true);
         example.setIsJetpack(true);
         example.setIsVisible(true);
-        example.setXmlRpcUrl("http://jetpack.url/xmlrpc.php");
+        example.setXmlRpcUrl("http://jetpack2.url/xmlrpc.php");
+        return example;
+    }
+
+    public static SiteModel generateSelfHostedSiteFutureJetpack() {
+        SiteModel example = new SiteModel();
+        example.setSelfHostedSiteId(8);
+        example.setIsWPCom(false);
+        example.setIsJetpack(false);
+        example.setIsVisible(true);
+        example.setXmlRpcUrl("http://jetpack2.url/xmlrpc.php");
         return example;
     }
 

--- a/example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java
+++ b/example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java
@@ -25,11 +25,11 @@ public class SiteUtils {
         return example;
     }
 
-    public static SiteModel generateJetpackSite() {
+    public static SiteModel generateJetpackSiteOverXMLRPC() {
         SiteModel example = new SiteModel();
         example.setSiteId(982);
         example.setSelfHostedSiteId(8);
-        example.setIsWPCom(true);
+        example.setIsWPCom(false);
         example.setIsJetpack(true);
         example.setIsVisible(true);
         example.setXmlRpcUrl("http://jetpack.url/xmlrpc.php");

--- a/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
@@ -79,7 +79,7 @@ public class SiteSqlUtils {
         if (siteResult.isEmpty()) {
             // No site with this local ID, or REMOTE_ID + URL, then insert it
             WellSql.insert(site).asSingleTransaction(true).execute();
-            return 0;
+            return 1;
         } else {
             // Update old site
             int oldId = siteResult.get(0).getId();

--- a/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
@@ -14,6 +14,9 @@ import org.wordpress.android.fluxc.model.SiteModel;
 import java.util.List;
 
 public class SiteSqlUtils {
+    public static class DuplicateSiteException extends Exception {
+    }
+
     public static List<SiteModel> getAllSitesWith(String field, Object value) {
         return WellSql.select(SiteModel.class)
                 .where().equals(field, value).endWhere()
@@ -55,7 +58,7 @@ public class SiteSqlUtils {
                 .getAsCursor().getCount();
     }
 
-    public static int insertOrUpdateSite(SiteModel site) {
+    public static int insertOrUpdateSite(SiteModel site) throws DuplicateSiteException {
         if (site == null) {
             return 0;
         }
@@ -74,6 +77,18 @@ public class SiteSqlUtils {
                     .equals(SiteModelTable.SITE_ID, site.getSiteId())
                     .equals(SiteModelTable.URL, site.getUrl())
                     .endGroup().endWhere().getAsModel();
+        }
+
+        // If the site is a self hosted, maybe it's already in the DB as a Jetpack site, and we don't want to create
+        // a duplicate.
+        if (siteResult.isEmpty() && !site.isWPCom()) {
+            siteResult = WellSql.select(SiteModel.class)
+                    .where()
+                    .equals(SiteModelTable.XMLRPC_URL, site.getXmlRpcUrl())
+                    .endWhere().getAsModel();
+            if (!siteResult.isEmpty()) {
+                throw new DuplicateSiteException();
+            }
         }
 
         if (siteResult.isEmpty()) {

--- a/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
@@ -77,7 +77,6 @@ public class SiteSqlUtils {
         // Looks like a new site, make sure we don't already have it.
         if (siteResult.isEmpty()) {
             AppLog.d(T.DB, "Site not found using local id");
-            // TODO: Make the URL enough, we could get surprise with the site id with .org sites becoming jetpack sites
             siteResult = WellSql.select(SiteModel.class)
                     .where().beginGroup()
                     .equals(SiteModelTable.SITE_ID, site.getSiteId())

--- a/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
@@ -81,13 +81,19 @@ public class SiteSqlUtils {
 
         // If the site is a self hosted, maybe it's already in the DB as a Jetpack site, and we don't want to create
         // a duplicate.
-        if (siteResult.isEmpty() && !site.isWPCom()) {
+        if (siteResult.isEmpty()) {
             siteResult = WellSql.select(SiteModel.class)
                     .where()
                     .equals(SiteModelTable.XMLRPC_URL, site.getXmlRpcUrl())
                     .endWhere().getAsModel();
             if (!siteResult.isEmpty()) {
-                throw new DuplicateSiteException();
+                // If the site already in the DB is a self hosted and the new one is a .com, it means we upgraded from
+                // self hosted to jetpack, we want to update the site with the new informations.
+                if (siteResult.get(0).isWPCom() || !site.isWPCom()) {
+                    // In other cases (examples: adding the same self hosted twice or adding self hosted on top of an
+                    // existing jetpack site), we consider it as an error.
+                    throw new DuplicateSiteException();
+                }
             }
         }
 

--- a/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java
@@ -10,6 +10,8 @@ import com.yarolegovich.wellsql.mapper.InsertMapper;
 
 import org.wordpress.android.fluxc.model.PostFormatModel;
 import org.wordpress.android.fluxc.model.SiteModel;
+import org.wordpress.android.util.AppLog;
+import org.wordpress.android.util.AppLog.T;
 
 import java.util.List;
 
@@ -63,14 +65,18 @@ public class SiteSqlUtils {
             return 0;
         }
 
+        AppLog.d(T.DB, "insertOrUpdateSite: " + site.getUrl());
+
         // If the site already exist and has an id, we want to update it.
         List<SiteModel> siteResult = WellSql.select(SiteModel.class)
                 .where().beginGroup()
                 .equals(SiteModelTable.ID, site.getId())
                 .endGroup().endWhere().getAsModel();
 
+
         // Looks like a new site, make sure we don't already have it.
         if (siteResult.isEmpty()) {
+            AppLog.d(T.DB, "Site not found using local id");
             // TODO: Make the URL enough, we could get surprise with the site id with .org sites becoming jetpack sites
             siteResult = WellSql.select(SiteModel.class)
                     .where().beginGroup()
@@ -79,17 +85,21 @@ public class SiteSqlUtils {
                     .endGroup().endWhere().getAsModel();
         }
 
+
         // If the site is a self hosted, maybe it's already in the DB as a Jetpack site, and we don't want to create
         // a duplicate.
         if (siteResult.isEmpty()) {
+            AppLog.d(T.DB, "Site not found using remote id + url");
             siteResult = WellSql.select(SiteModel.class)
                     .where()
                     .equals(SiteModelTable.XMLRPC_URL, site.getXmlRpcUrl())
                     .endWhere().getAsModel();
             if (!siteResult.isEmpty()) {
+                AppLog.d(T.DB, "Site found using xmlrpc url: " + site.getXmlRpcUrl());
                 // If the site already in the DB is a self hosted and the new one is a .com, it means we upgraded from
                 // self hosted to jetpack, we want to update the site with the new informations.
                 if (siteResult.get(0).isWPCom() || !site.isWPCom()) {
+                    AppLog.d(T.DB, "Site is a duplicate");
                     // In other cases (examples: adding the same self hosted twice or adding self hosted on top of an
                     // existing jetpack site), we consider it as an error.
                     throw new DuplicateSiteException();
@@ -98,11 +108,13 @@ public class SiteSqlUtils {
         }
 
         if (siteResult.isEmpty()) {
-            // No site with this local ID, or REMOTE_ID + URL, then insert it
+            // No site with this local ID, REMOTE_ID + URL, or XMLRPC URL, then insert it
+            AppLog.d(T.DB, "Inserting site: " + site.getUrl());
             WellSql.insert(site).asSingleTransaction(true).execute();
             return 1;
         } else {
             // Update old site
+            AppLog.d(T.DB, "Updating site: " + site.getUrl());
             int oldId = siteResult.get(0).getId();
             return WellSql.update(SiteModel.class).whereId(oldId)
                     .put(site, new UpdateAllExceptId<SiteModel>()).execute();

--- a/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.java
@@ -25,6 +25,7 @@ import org.wordpress.android.fluxc.network.rest.wpcom.site.SiteRestClient.IsWPCo
 import org.wordpress.android.fluxc.network.rest.wpcom.site.SiteRestClient.NewSiteResponsePayload;
 import org.wordpress.android.fluxc.network.xmlrpc.site.SiteXMLRPCClient;
 import org.wordpress.android.fluxc.persistence.SiteSqlUtils;
+import org.wordpress.android.fluxc.persistence.SiteSqlUtils.DuplicateSiteException;
 import org.wordpress.android.util.AppLog;
 import org.wordpress.android.util.AppLog.T;
 
@@ -165,6 +166,7 @@ public class SiteStore extends Store {
 
     public enum SiteErrorType {
         INVALID_SITE,
+        DUPLICATE_SITE,
         GENERIC_ERROR
     }
 
@@ -697,27 +699,31 @@ public class SiteStore extends Store {
     }
 
     private void updateSite(SiteModel siteModel) {
-        OnSiteChanged event;
+        OnSiteChanged event = new OnSiteChanged(0);
         if (siteModel.isError()) {
-            event = new OnSiteChanged(0);
             // TODO: what kind of error could we get here?
             event.error = new SiteError(SiteErrorType.GENERIC_ERROR);
         } else {
-            int rowsAffected = SiteSqlUtils.insertOrUpdateSite(siteModel);
-            event = new OnSiteChanged(rowsAffected);
+            try {
+                event.rowsAffected = SiteSqlUtils.insertOrUpdateSite(siteModel);
+            } catch (DuplicateSiteException e) {
+                event.error = new SiteError(SiteErrorType.DUPLICATE_SITE);
+            }
         }
         emitChange(event);
     }
 
     private void updateSites(SitesModel sitesModel) {
-        OnSiteChanged event;
+        OnSiteChanged event = new OnSiteChanged(0);
         if (sitesModel.isError()) {
-            event = new OnSiteChanged(0);
             // TODO: what kind of error could we get here?
             event.error = new SiteError(SiteErrorType.GENERIC_ERROR);
         } else {
-            int rowsAffected = createOrUpdateSites(sitesModel);
-            event = new OnSiteChanged(rowsAffected);
+            try {
+                event.rowsAffected = createOrUpdateSites(sitesModel);
+            } catch (DuplicateSiteException e) {
+                event.error = new SiteError(SiteErrorType.DUPLICATE_SITE);
+            }
         }
         emitChange(event);
     }
@@ -765,7 +771,7 @@ public class SiteStore extends Store {
         emitChange(event);
     }
 
-    private int createOrUpdateSites(SitesModel sites) {
+    private int createOrUpdateSites(SitesModel sites) throws DuplicateSiteException {
         int rowsAffected = 0;
         for (SiteModel site : sites.getSites()) {
             rowsAffected += SiteSqlUtils.insertOrUpdateSite(site);

--- a/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.java
@@ -757,7 +757,7 @@ public class SiteStore extends Store {
     private void handleCheckedIsWPComUrl(IsWPComResponsePayload payload) {
         OnURLChecked event = new OnURLChecked(payload.url);
         if (payload.isError()) {
-            // Return invalid site for all errors (this endpoint is not documented and seems a bit drunk).
+            // Return invalid site for all errors (this endpoint seems a bit drunk).
             // Client likely needs to know if there was an error or not.
             event.error = new SiteError(SiteErrorType.INVALID_SITE);
         }

